lib/gen_fsm.ex

```diff
@@ -20,69 +85,82 @@ defmodule GenFSM do
     end
   end
 
+  @spec stop(server, reason :: term, timeout) :: :ok
   def stop(fsm, reason \\ :normal, timeout \\ :infinity) do
     :gen.stop(fsm, reason, timeout)
   end
 
+  @type fsm_ref :: pid | name
+
+  @spec send_event(fsm_ref, event) :: :ok
   def send_event(fsm, event) do
     :gen_fsm.send_event(fsm, event)
   end
 
+  @spec send_all_state_event(fsm_ref, event) :: :ok
   def send_all_state_event(fsm, event) do
     :gen_fsm.send_all_state_event(fsm, event)
   end
 
+  @spec sync_send_event(fsm_ref, event, timeout) :: :ok
   def sync_send_event(fsm, event, timeout \\ 5000) do
     :gen_fsm.sync_send_event(fsm, event, timeout)
   end
 
+  @spec sync_send_all_state_event(fsm_ref, event, timeout) :: :ok
   def sync_send_all_state_event(fsm, event, timeout \\ 5000) do
     :gen_fsm.sync_send_all_state_event(fsm, event, timeout)
   end
 
+  @type from :: {pid, tag :: term}
+  @spec reply(from, term) :: :ok
   def reply(caller, reply) do
     :gen_fsm.reply(caller, reply)
   end
 
+  @spec send_event_after(integer, event) :: reference
   def send_event_after(time, event) do
     :gen_fsm.send_event_after(time, event)
   end
 
+  @spec start_timer(integer, term) :: reference
   def start_timer(time, message) do
     :gen_fsm.start_timer(time, message)
   end
 
+  @type remaining_time :: integer
+  @spec cancel_timer(reference) :: remaining_time | false
   def cancel_timer(timer_ref) do
     :gen_fsm.cancel_timer(timer_ref)
   end
 
   defmacro __using__(_) do
     quote location: :keep do
-     @behaviour :gen_fsm
+      @behaviour :gen_fsm
 
       @doc false
-      def handle_event(event, state_name, state_data) do
-        { :stop, {:bad_event, state_name, event}, state_data }
+      def handle_event(_event, _state_name, state_data) do
+        {:stop, :unexpected_event, state_data}
```

Comment:
wouldn't it be helpful to return the state_name and event in case of an unexpected event?  That would make it easier to debug?  Or is there a good reason you removed it?